dynamic_programming/subset_generation.py

```diff
@@ -9,16 +9,32 @@ def combination_util(arr, n, r, index, data, i):
     start & end ---> Staring and Ending indexes in arr[]
     index  ---> Current index in data[]
     r ---> Size of a combination to be printed
+
+    >>> combination_util([1, 2, 3], 3, 2, 0, [0, 0], 0)
+    1 2
+    1 3
+    2 3
+    >>> combination_util([-1,-2,-3], 3, 2, 0, [0, 0], 0)
+    -1 -2
+    -1 -3
+    -2 -3
+
+    >>> combination_util([], 0, 2, 0, [0, 0], 0)
+
+    >>> combination_util([1], 1, 2, 0, [0, 0], 0)
+
+    >>> combination_util([1], 1, 2, 0, [0, 0], 3)
+
+
     """
     if index == r:
-        for j in range(r):
-            print(data[j], end=" ")
-        print(" ")
+        print(" ".join(map(str, data[:r])))
```

Comment:
CONTRIBUTING.md says that algorithmic functions should not `print()` instead they should return or yield values or raise exceptions.